File: app/src/lib/types/performance.ts
/** Return, risk and benchmark over one period. */

import type { GrowthSeries, StatSeries, ValueSeries } from "./periods";
import type { DateString, MoneyString } from "./primitives";
import type { ChargeSummary } from "./reports";
import type { TradingVolume } from "./transactions";
export interface PeriodReturn {
  from: DateString;
  to: DateString;
  twr: MoneyString;
}

/** The money side of a period, next to the return side a TWR gives. */
export interface PeriodSummary {
  start_value_base: MoneyString;
  end_value_base: MoneyString;
  /** Net external flow; `+` is money brought in. */
  net_flow_base: MoneyString;
  /** End minus start: the change a statement shows, deposits included. */
  absolute_change_base: MoneyString;
  /** The change with the flows taken out — what was earned. */
  delta_base: MoneyString;
  /** Money committed by the end of the period. */
  invested_capital_base: MoneyString;
  /** Capital weighted by the time it was invested; the denominator of every cost rate. */
  average_capital_base: MoneyString;
}

export interface PerformanceData {
  from: DateString;
  to: DateString;
  base_currency: string;
  twr: MoneyString;
  /** `twr` as a yearly rate; `null` for a period shorter than a day. */
  twr_annualized: MoneyString | null;
  xirr: MoneyString | null;
  summary: PeriodSummary;
  /** Paid in the period, trade commissions included — unlike the charges report. */
  costs: ChargeSummary;
  fee_rate: MoneyString | null;
  tax_rate: MoneyString | null;
  /** Bought and sold in the period, and that volume over the capital at work. */
  volume: TradingVolume;
  turnover_rate: MoneyString | null;
  /** The period's highest value; `null` for a portfolio that never held anything. */
  peak: Peak | null;
  series: ValueSeries;
  growth: GrowthSeries;
  monthly_returns: PeriodReturn[];
  annual_returns: PeriodReturn[];
}

/** Calendar granularity the calculation sheet is split by. */
export type SheetPeriod = "DAY" | "WEEK" | "MONTH" | "QUARTER" | "YEAR";

/** One chunk of the calculation sheet: `end = start + flow + delta`, and what made the delta. */
export interface CalculationRow {
  from: DateString;
  to: DateString;
  /** The previous row's close; for the first row, the value before the window opened. */
  start_value_base: MoneyString;
  external_flow_base: MoneyString;
  /** Dividends and interest, gross — the tax withheld from them sits in `costs`. */
  income_base: MoneyString;
  costs: ChargeSummary;
  /** Fees and taxes added up, summed in the host so the UI never adds money strings. */
  costs_base: MoneyString;
  /** What is left of the change once income and costs are named. */
  market_change_base: MoneyString;
  delta_base: MoneyString;
  end_value_base: MoneyString;
  twr: MoneyString;
  /** The chunks so far, chained; the last row's value is the period's own return. */
  cumulative_twr: MoneyString;
}

export interface CalculationSheet {
  base_currency: string;
  period: SheetPeriod;
  rows: CalculationRow[];
  total: PeriodSummary;
  twr: MoneyString;
}

export interface Drawdown {
  peak: DateString;
  trough: DateString;
  recovered: DateString | null;
  /** Drawdown fraction; `-0.35` means -35%. */
  depth: number;
}

export interface RiskMetrics {
  days: number;
  volatility: number;
  semi_deviation: number;
  max_drawdown: Drawdown | null;
  /** Calendar days the deepest episode lasted, counted to the period end while underwater. */
  max_drawdown_days: number | null;
  /** Longest episode by duration, which is rarely the deepest one. */
  longest_drawdown: Drawdown | null;
  longest_drawdown_days: number | null;
  /** Distance from the peak right now; `0` means the period ends at one. */
  current_drawdown: number;
  /** The peak not yet climbed back to. */
  current_drawdown_since: DateString | null;
  sharpe: number | null;
  annualized_return: number;
  best_day: [DateString, number] | null;
  worst_day: [DateString, number] | null;
  positive_days_share: number;
}

/** Complete payload for the risk screen. */
export interface RiskReport {
  metrics: RiskMetrics;
  window_days: number;
  returns: StatSeries;
  rolling_volatility: StatSeries;
  drawdown: StatSeries;
  episodes: Drawdown[];
}

export interface BenchmarkComparison {
  from: DateString;
  to: DateString;
  portfolio_twr: MoneyString;
  benchmark_twr: MoneyString;
  excess: MoneyString;
}

// Allocation, taxonomies, and targets

/** Highest point of a series and where the last point stands against it. */
export interface Peak {
  date: DateString;
  value: MoneyString;
  current: MoneyString;
  /** `current / peak - 1`: zero at the high, negative below it. */
  distance: MoneyString | null;
  days_since: number;
}

/** A nominal return restated in the money of the period's first day. */
export interface RealReturn {
  /** ISO 3166-1 alpha-2, or `EA` / `EU` for the aggregates. The name is written here. */
  region: string;
  from: DateString;
  /** Where the index stops, which may be earlier than the period asked for. */
  to: DateString;
  /** What one unit of money at `from` costs at `to`: `1.04` is 4% of inflation. */
  factor: MoneyString;
  /** The same thing as a rate — `factor - 1`, computed by the host. */
  inflation: MoneyString;
  nominal: MoneyString;
  real: MoneyString;
  real_annualized: MoneyString | null;
}

/** A period's returns with inflation taken out; absent when no region is set. */
export interface RealPerformance {
  twr: RealReturn;
  /** Money-weighted, each flow deflated at its own date. */
  xirr: MoneyString | null;
}

/** What the app knows about the portfolio's price index. Codes only, never country names. */
export interface InflationStatus {
  region: string | null;
  source: string | null;
  /** First day of the last month a level is stored for. */
  published_through: DateString | null;
  /** Every region this build can fetch, for the picker. */
  regions: string[];
}

File: app/src-tauri/src/commands/performance.rs
use crate::commands::named;
use crate::commands::parse_date;
use crate::error::{UiError, UiResult};
use crate::scope::DataScope;
use crate::state::AppState;
use rust_decimal::Decimal;
use serde::Serialize;
use sq_core::calc::{
    CalculationSheet, ChargeSummary, GrowthSeries, Peak, Period, PeriodReturn, PeriodSummary, RiskReport,
    TradingVolume, ValueSeries, all_time_high,
};
use sq_core::market::DateRange;
use tauri::State;

#[derive(Debug, Serialize)]
pub struct PerformanceData {
    pub from: String,
    pub to: String,
    pub base_currency: String,
    #[serde(with = "rust_decimal::serde::str")]
    pub twr: Decimal,
    /// The same return as a yearly rate; `None` for a period shorter than a day.
    #[serde(with = "rust_decimal::serde::str_option")]
    pub twr_annualized: Option<Decimal>,
    #[serde(with = "rust_decimal::serde::str_option")]
    pub xirr: Option<Decimal>,
    /// Values, flows, and the capital they add up to — the money side of the same period.
    pub summary: PeriodSummary,
    /// Fees and taxes paid in the period, trade commissions included.
    pub costs: ChargeSummary,
    #[serde(with = "rust_decimal::serde::str_option")]
    pub fee_rate: Option<Decimal>,
    #[serde(with = "rust_decimal::serde::str_option")]
    pub tax_rate: Option<Decimal>,
    /// Bought and sold in the period, and that volume against the capital at work.
    pub volume: TradingVolume,
    #[serde(with = "rust_decimal::serde::str_option")]
    pub turnover_rate: Option<Decimal>,
    /// The period's highest value and how far below it the period ends; `None` for an
    /// empty portfolio, which has no high to stand under.
    pub peak: Option<Peak>,
    pub series: ValueSeries,
    pub growth: GrowthSeries,
    pub monthly_returns: Vec<PeriodReturn>,
    pub annual_returns: Vec<PeriodReturn>,
}

#[tauri::command]
pub fn performance_summary(
    state: State<AppState>,
    from: String,
    to: String,
    source: Option<DataScope>,
) -> UiResult<PerformanceData> {
    let range = date_range(&from, &to)?;
    let store = state.store()?;
    let scope = state.scope_selection_in(&store, source.as_ref())?;
    let analytics = scope.analytics(&store)?;

    let series = analytics.series(range).map_err(|e| named(&store, e))?;
    let growth = series.growth();
    let monthly_returns = sq_core::calc::returns_by_period(&series, Period::Month);
    let annual_returns = sq_core::calc::returns_by_period(&series, Period::Year);

    let summary = sq_core::calc::period_summary(&series);
    let costs = analytics
        .costs(range.from, range.to)
        .map_err(|e| named(&store, e))?;
    let twr = analytics.twr(range.from, range.to)?;
    let volume = analytics.trading_volume(range.from, range.to)?;

    Ok(PerformanceData {
        from: range.from.to_string(),
        to: range.to.to_string(),
        base_currency: analytics.base_currency().to_string(),
        twr,
        twr_annualized: sq_core::calc::annualize(twr, range.from, range.to),
        xirr: match analytics.xirr(range.to) {
            Ok(v) => Some(v),
            Err(sq_core::Error::Math(_)) => None,
            Err(e) => return Err(UiError::from(e)),
        },
        fee_rate: summary.rate_of(costs.fees_base),
        tax_rate: summary.rate_of(costs.taxes_base),
        turnover_rate: summary.rate_of(volume.volume_base),
        volume,
        peak: all_time_high(&series),
        summary,
        costs,
        growth,
        monthly_returns,
        annual_returns,
        series,
    })
}

/// The calculation sheet of the period: one row per calendar chunk, showing how the opening
/// value, the flows and what was earned add up to the closing one. The granularity is the
/// screen's, not the core's — a decade reads by year and a month by day.
#[tauri::command]
pub fn performance_breakdown(
    state: State<AppState>,
    from: String,
    to: String,
    period: Period,
    source: Option<DataScope>,
) -> UiResult<CalculationSheet> {
    let range = date_range(&from, &to)?;
    let store = state.store()?;
    let scope = state.scope_selection_in(&store, source.as_ref())?;
    scope
        .analytics(&store)?
        .calculation_sheet(range, period)
        .map_err(|e| named(&store, e))
}

/// The calculation sheet as a CSV of what the panel shows, written where the user points.
#[tauri::command]
pub fn performance_sheet_save(
    state: State<AppState>,
    from: String,
    to: String,
    period: Period,
    path: String,
) -> UiResult<()> {
    use crate::commands::reports::csv::{money, row};

    let sheet = performance_breakdown(state, from, to, period, None)?;
    let c = &sheet.base_currency;
    let mut csv = row([
        "from",
        "to",
        &format!("opening {c}"),
        &format!("flows {c}"),
        &format!("market {c}"),
        &format!("income {c}"),
        &format!("costs {c}"),
        &format!("result {c}"),
        &format!("closing {c}"),
        "return",
        "chained return",
    ]);
    for r in &sheet.rows {
        csv.push_str(&row([
            r.from.to_string().as_str(),
            r.to.to_string().as_str(),
            money(r.start_value_base).as_str(),
            money(r.external_flow_base).as_str(),
            money(r.market_change_base).as_str(),
            money(r.income_base).as_str(),
            money(r.costs_base).as_str(),
            money(r.delta_base).as_str(),
            money(r.end_value_base).as_str(),
            money(r.twr).as_str(),
            money(r.cumulative_twr).as_str(),
        ]));
    }
    csv.push_str(&row([
        sheet
            .rows
            .first()
            .map(|r| r.from.to_string())
            .unwrap_or_default()
            .as_str(),
        sheet
            .rows
            .last()
            .map(|r| r.to.to_string())
            .unwrap_or_default()
            .as_str(),
        money(sheet.total.start_value_base).as_str(),
        money(sheet.total.net_flow_base).as_str(),
        "",
        "",
        "",
        money(sheet.total.delta_base).as_str(),
        money(sheet.total.end_value_base).as_str(),
        "",
        money(sheet.twr).as_str(),
    ]));

    // Excel reads a semicolon-separated file as UTF-8 only when it opens with a BOM.
    let mut bytes = vec![0xEF, 0xBB, 0xBF];
    bytes.extend_from_slice(csv.as_bytes());
    std::fs::write(&path, bytes).map_err(|e| UiError::invalid(format!("cannot write {path}: {e}")))
}

#[tauri::command]
pub fn risk_report(
    state: State<AppState>,
    from: String,
    to: String,
    risk_free_rate: f64,
    window_days: usize,
    source: Option<DataScope>,
) -> UiResult<RiskReport> {
    let range = date_range(&from, &to)?;
    let store = state.store()?;
    let scope = state.scope_selection_in(&store, source.as_ref())?;
    scope
        .analytics(&store)?
        .risk_report(range, risk_free_rate, window_days)
        .map_err(|e| named(&store, e))
}

#[tauri::command]
pub fn benchmark_compare(
    state: State<AppState>,
    security_id: String,
    from: String,
    to: String,
    source: Option<DataScope>,
) -> UiResult<sq_core::calc::BenchmarkComparison> {
    let range = date_range(&from, &to)?;
    let store = state.store()?;
    let scope = state.scope_selection_in(&store, source.as_ref())?;
    scope
        .analytics(&store)?
        .benchmark(&security_id, range.from, range.to)
        .map_err(|e| named(&store, e))
}

#[tauri::command]
pub fn benchmark_series(
    state: State<AppState>,
    security_id: String,
    from: String,
    to: String,
    source: Option<DataScope>,
) -> UiResult<GrowthSeries> {
    let range = date_range(&from, &to)?;
    let store = state.store()?;
    let scope = state.scope_selection_in(&store, source.as_ref())?;
    let dates: Vec<chrono::NaiveDate> = std::iter::successors(Some(range.from), |d| {
        d.succ_opt().filter(|next| *next <= range.to)
    })
    .collect();
    scope
        .analytics(&store)?
        .benchmark_growth(&security_id, &dates)
        .map_err(|e| named(&store, e))
}

pub(crate) fn date_range(from: &str, to: &str) -> UiResult<DateRange> {
    let from = parse_date(from)?;
    let to = parse_date(to)?;
    if to < from {
        return Err(UiError::invalid("the period ends before it starts"));
    }
    Ok(DateRange::new(from, to))
}
